server/src/main/java/org/elasticsearch/action/bulk/TransportAbstractBulkAction.java

```diff
@@ -228,10 +228,18 @@ private boolean applyPipelines(Task task, BulkRequest bulkRequest, Executor exec
             metadata = clusterService.state().getMetadata();
         }
 
+        Map<String, IngestService.Pipelines> resolvedPipelineCache = new HashMap<>();
         for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
             IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
             if (indexRequest != null) {
-                IngestService.resolvePipelinesAndUpdateIndexRequest(actionRequest, indexRequest, metadata);
```

Comment:
I am been having trouble unit testing this bit of code that does caching because it's inlined in the loop. I'm not super worried, because it is very similar to ` resolvePipelinesAndUpdateIndexRequest()` which is tested well (but doesn't do the caching).

We discussed previously (and discarded) some ideas for encapsulating this logic so it could be tested. Eg, having `resolvePipelinesAndUpdateIndexRequest` take a reference to `resolvedPipelineCache`, but that was decidedly nasty.

The only other idea I have, it to encapsulate `resolvedPipelineCache` in an class called `CachedPipelineResolver` with a function `resolvePipelinesAndUpdateIndexRequest` consisting of the lines of logic below.

Thoughts?

Reply:
This is what I meant:
```
   public static class CachedPipelineResolver {
        private final Map<String, Pipelines> resolvedPipelineCache;

        public CachedPipelineResolver() {
            this(new HashMap<>());
        }

        // For testing
        CachedPipelineResolver(Map<String, Pipelines> resolvedPipelineCache) {
            this.resolvedPipelineCache = resolvedPipelineCache;
        }

        public void resolvePipelinesAndUpdateIndexRequest(
            final DocWriteRequest<?> originalRequest,
            final IndexRequest indexRequest,
            final Metadata metadata
        ) {
            if (indexRequest.isPipelineResolved() == false) {
                var pipeline = resolvedPipelineCache.computeIfAbsent(
                    indexRequest.index(),
                    // TODO perhaps this should use `threadPool.absoluteTimeInMillis()`, but leaving as is for now.
                    (index) -> IngestService.resolveStoredPipelines(originalRequest, indexRequest, metadata, System.currentTimeMillis())
                );
                IngestService.setPipelineOnRequest(indexRequest, pipeline);
            }
        }
    
```

Reply:
If y'all like this approach, I have an un-pushed commit with this change and more unit tests.

Reply:
I don't have strong feelings either way.

Reply:
I'm in favor of that version of the code (or something not entirely unlike it), and I think it resolves my issues about how intimate these two classes are getting with each other. Given the plan for backporting this PR, I think we should punt on that change until a subsequent PR.

That is, let's merge this PR more or less as-is to all the branches, but then we can follow up with a separate refactoring PR for `main` and `8.x` only next week.